### PR TITLE
Add hcloud_floating_ip module

### PR DESCRIPTION
**Pull request by LKaemmerling**
##### SUMMARY

This is now the completion of all Hetzner Cloud related modules. 

The first attempt was performed a long time ago for Ansible 2.8 (#53551). While developing it we noticed a small inconsistency within our resources - the Floating IPs was missing a user-controlled unique property - `name`. We have this now implemented in our system, so now we are able to contribute finally a `hcloud_floating_ip` Module to Ansible 2.10 🎉 

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

hcloud_floating_ip
##### ADDITIONAL INFORMATION

```paste below

```
//cc @resmo

### CODE REVIEW

**Comment by resmo** [comment]
Very good decision adding a name! love it :heart:

File: lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py [hunk 1]
```diff
@@ -0,0 +1,303 @@
...
+                force_assign={"type": "bool"},
+                type={"choices": ["ipv4", "ipv6"], "required": True},
+                labels={"type": "dict"},
+                state={
+                    "choices": ["absent", "present"],
+                    "default": "present",
+                },
+                **Hcloud.base_module_arguments()
+            ),
+            required_one_of=[['home_location', 'server'], ['id', 'name']],
```
Comment (resmo) [inline lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py]: AFAICS is home_location or server only required on "creation" right? and there is already a check for this. So change to this?

~~~diff
-            required_one_of=[['home_location', 'server'], ['id', 'name']],
+            required_one_of=[['id', 'name']],
~~~

File: lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py [hunk 2]
```diff
@@ -0,0 +1,303 @@
...
+        description = self.module.params.get("description")
+        if description is not None and description != self.hcloud_floating_ip.description:
+            if not self.module.check_mode:
+                self.hcloud_floating_ip.update(description=description)
+            self._mark_as_changed()
+
+        server = self.module.params.get("server")
+        if server is not None:
+            if not self.module.check_mode:
+                if self.module.params.get("force_assign") or self.hcloud_floating_ip.server is None:
```
Comment (resmo) [inline lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py]: could this check go outside the `if checkmode` conditio,  so a user get the same consistent "error" regardless if check mode or not?

File: lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py [hunk 3]
```diff
@@ -0,0 +1,303 @@
...
+        if self.hcloud_floating_ip is None:
+            self._create_floating_ip()
+        else:
+            self._update_floating_ip()
+
+    def delete_floating_ip(self):
+        self._get_floating_ip()
+        if self.hcloud_floating_ip is not None:
+            if not self.module.check_mode:
+                self.client.floating_ips.delete(self.hcloud_floating_ip)
```
Comment (resmo) [inline lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py]: probably worth considering is a check if floating IP is assigned and warn (same as in the update case). A `force=yes` (probably consolidate force_assign to force) could delete an assigned IP.

File: lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py [hunk 4]
```diff
@@ -0,0 +1,303 @@
...
+    def define_module():
+        return AnsibleModule(
+            argument_spec=dict(
+                id={"type": "int"},
+                name={"type": "str"},
+                description={"type": "str"},
+                server={"type": "str"},
+                home_location={"type": "str"},
+                force_assign={"type": "bool"},
+                type={"choices": ["ipv4", "ipv6"], "required": True},
```
Comment (resmo) [inline lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py]: I would suggest to default this to `ipv4` or alternatively make it required only if `state=present`. because it would be weird to delete an e.g. ipv4 by name and specify type=ipv6 right?

File: lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py [hunk 5]
```diff
@@ -0,0 +1,303 @@
...
+  hcloud_floating_ip:
+    name: my-floating-ip
+    home_location: fsn1
+    type: ipv4
+    state: present
+
+- name: Ensure the Network is absent (remove if needed)
+  hcloud_floating_ip:
+    name: my-network
+    state: absent
```
Comment (resmo) [inline lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py:110]: currently this example does not work but it would be nice it would this way

File: lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py [hunk 6]
```diff
@@ -0,0 +1,328 @@
...
+        else:
+            self._update_floating_ip()
+
+    def delete_floating_ip(self):
+        self._get_floating_ip()
+        if self.hcloud_floating_ip is not None:
+            if self.module.params.get("force") or self.hcloud_floating_ip.server is None:
+                if not self.module.check_mode:
+                    self.client.floating_ips.delete(self.hcloud_floating_ip)
+                else:
```
Comment (resmo) [inline lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py]: indentation looks wrong. else should be the outer if

**Review by goneri — CHANGES_REQUESTED** [review]
Could you please add `closes: #62414` in the commit message and remove the `disabled` from `test/integration/targets/hcloud_floating_ip_info/aliases`

File: lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py [hunk 7]
```diff
@@ -0,0 +1,328 @@
...
+- name: Assign a Floating IP to a server
+  hcloud_floating_ip:
+    name: my-floating-ip
+    server: 1234
+    state: present
+- name: Assign a Floating IP to another server
+  hcloud_floating_ip:
+    name: my-floating-ip
+    server: 1234
+    force_assign: yes
```
Comment (resmo) [inline lib/ansible/modules/cloud/hcloud/hcloud_floating_ip.py]: should be `force`

File: test/integration/targets/hcloud_floating_ip_info/aliases [hunk 8]
```diff
@@ -1,3 +1,2 @@
 cloud/hcloud
 shippable/hcloud/group1
-disabled # See: https://github.com/ansible/ansible/issues/62414
```
Comment (mattclay) [inline test/integration/targets/hcloud_floating_ip_info/aliases]: This should be done in a separate PR so it can be backported to the stable branches which also disabled the test.

File: test/integration/targets/hcloud_floating_ip/tasks/main.yml [hunk 9]
```diff
@@ -0,0 +1,276 @@
...
+    description: "changed-description"
+    type: ipv4
+    home_location: "fsn1"
+    labels:
+      key: value
+  register: floatingIP
+- name: verify test update Floating IP
+  assert:
+    that:
+      - floatingIP is success
```
Comment (resmo) [inline test/integration/targets/hcloud_floating_ip/tasks/main.yml]:
~~~diff
-  - floatingIP is success
+  - floatingIP is changed
~~~

File: test/integration/targets/hcloud_floating_ip/tasks/main.yml [hunk 10]
```diff
@@ -0,0 +1,276 @@
...
+    name: "{{ hcloud_floating_ip_name }}"
+    description: "Web Server"
+    type: ipv4
+    home_location: "fsn1"
+  register: floatingIP
+- name: verify test create Floating IP
+  assert:
+    that:
+      - floatingIP is success
+      - floatingIP.hcloud_floating_ip.name == "{{ hcloud_floating_ip_name }}"
```
Comment (resmo) [inline test/integration/targets/hcloud_floating_ip/tasks/main.yml]: not a blocker but this could be simplified:
~~~diff
-       - floatingIP.hcloud_floating_ip.name == "{{ hcloud_floating_ip_name }}"
+       - floatingIP.hcloud_floating_ip.name == hcloud_floating_ip_name
~~~

File: test/integration/targets/hcloud_floating_ip/tasks/main.yml [hunk 11]
```diff
@@ -0,0 +1,276 @@
...
+- name: test unassign Floating IP
+  hcloud_floating_ip:
+    name: "{{ hcloud_floating_ip_name }}"
+    type: ipv4
+    home_location: "fsn1"
+  register: floatingIP
+- name: verify test unassign Floating IP
+  assert:
+    that:
+      - floatingIP is success
```
Comment (resmo) [inline test/integration/targets/hcloud_floating_ip/tasks/main.yml]:
~~~diff
-  - floatingIP is success
+  - floatingIP is changed
~~~

File: test/integration/targets/hcloud_floating_ip/tasks/main.yml [hunk 12]
```diff
@@ -0,0 +1,276 @@
...
+- name: test assign Floating IP again
+  hcloud_floating_ip:
+    name: "{{ hcloud_floating_ip_name }}"
+    type: ipv4
+    server: "{{ main_server.hcloud_server.name }}"
+  register: floatingIP
+- name: verify test assign Floating IP again
+  assert:
+    that:
+      - floatingIP is success
```
Comment (resmo) [inline test/integration/targets/hcloud_floating_ip/tasks/main.yml]:
~~~diff
-  - floatingIP is success
+  - floatingIP is changed
~~~

**Review by resmo — APPROVED** [review]
This module is in a very good shape, tests could be a bit extended, I miss some check mode tests and some idempotency tests but I won't hold this back.

shipit

**Comment by cschmitt-hcloud** [comment]
shipit

**Comment by resmo** [comment]
shipit

**Comment by resmo** [comment]
shipit

**Comment by resmo** [comment]
Thanks!
